ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryMonitor.java

```diff
@@ -0,0 +1,228 @@
+    @Override
+    public long read() {
+      return System.currentTimeMillis();
+    }
+  };
+
+  private static final int SHUTDOWN_TIMEOUT_MS = 5000;
+
+  private final Ticker ticker;
+  private final long retryBackoffInitialMs;
+  private final long retryBackoffMaxMs;
+  private final KsqlEngine ksqlEngine;
+  private final ExecutorService executor;
+  private final Map<QueryId, RetryEvent> queriesRetries = new HashMap<>();
+
+  private volatile boolean closed = false;
+
+  public QueryMonitor(final KsqlConfig ksqlConfig, final KsqlEngine ksqlEngine) {
+    this(
+        ksqlEngine,
+        Executors.newSingleThreadExecutor(r -> new Thread(r, QueryMonitor.class.getName())),
+        ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_INITIAL_MS),
+        ksqlConfig.getLong(KsqlConfig.KSQL_QUERY_RETRY_BACKOFF_MAX_MS),
+        CURRENT_TIME_MILLIS_TICKER
+    );
+  }
+
+  @VisibleForTesting
+  QueryMonitor(
+      final KsqlEngine ksqlEngine,
+      final ExecutorService executor,
+      final long retryBackoffInitialMs,
+      final long retryBackoffMaxMs,
+      final Ticker ticker
+  ) {
+    this.retryBackoffInitialMs = retryBackoffInitialMs;
+    this.retryBackoffMaxMs = retryBackoffMaxMs;
+    this.ksqlEngine = ksqlEngine;
+    this.executor = executor;
+    this.ticker = ticker;
+  }
+
+  public void start() {
+    executor.execute(new Runner());
+    executor.shutdown();
+  }
+
+  @Override
+  public void close() {
+    closed = true;
+
+    try {
+      executor.awaitTermination(SHUTDOWN_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    } catch (final InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+  }
+
+  boolean isClosed() {
+    return closed;
+  }
+
+  void restartFailedQueries() {
+    // Collect a list of new queries in ERROR state
+    ksqlEngine.getPersistentQueries().stream()
+        .filter(query -> query.getState() == KafkaStreams.State.ERROR)
+        .filter(query -> !queriesRetries.containsKey(query.getQueryId()))
+        .map(QueryMetadata::getQueryId)
+        .forEach(queryId -> queriesRetries.put(queryId, newRetryEvent(queryId)));
+
+    // Restart queries that has passed the waiting timeout
+    final List<QueryId> deleteRetryEvents = new ArrayList<>();
+    queriesRetries.entrySet().stream()
+        .forEach(mapEntry -> {
+          final QueryId queryId = mapEntry.getKey();
+          final RetryEvent retryEvent = mapEntry.getValue();
+          final Optional<PersistentQueryMetadata> query = ksqlEngine.getPersistentQuery(queryId);
+
+          // Query was terminated manually if no present
+          if (!query.isPresent()) {
+            deleteRetryEvents.add(queryId);
+          } else if (ticker.read() > retryEvent.nextRestartTimeMs()) {
+            if (query.get().getState() == KafkaStreams.State.ERROR) {
+              // Retry again if it's still in ERROR state
+              retryEvent.restart();
+            } else {
```

Comment:
I don't understand the `queryHealthyTime` part. The `QueryMonitor` isn't stamping queries as healthy. It is just restarting them. Or do you want to keep the `queryErrors` and `numRetries` during the `queryHealthyTime` in case the query is in ERROR state again after the backoff period has passed but before the query healthy time? 

Btw, I saw in the query error handling docs and it says that we might change the state of the query to RUNNING once it has passed the healthy time. Can we change this logic once that's implemented?

Reply:
> Can we change this logic once that's implemented?

Sure - I was just suggesting a minor refactor to make the distinction between the thresholds more clear. So instead of:

```
         if (!query.isPresent()) {
            deleteRetryEvents.add(queryId);
          } else if (ticker.read() > retryEvent.nextRestartTimeMs() && query.get().getState() == KafkaStreams.State.ERROR) {
              // Retry again if it's still in ERROR state
              retryEvent.restart();
           } else if (ticker.read() > retryEvent.queryHealthyTime()) {
                query.ifPresent(QueryMetadata::clearErrors);
               deleteRetryEvents.add(queryId);
           }
```

where `queryHealthyTime` just returns the same value as `nextRestartTimeMs` for now. Just a suggestion.